consensus/istanbul/backend/engine.go

```diff
@@ -95,15 +96,68 @@ var (
 	nonceAuthVote = hexutil.MustDecode("0xffffffffffffffff") // Magic nonce number to vote on adding a new validator
 	nonceDropVote = hexutil.MustDecode("0x0000000000000000") // Magic nonce number to vote on removing a validator.
 
-	inmemoryAddresses  = 20 // Number of recent addresses from ecrecover
-	recentAddresses, _ = lru.NewARC(inmemoryAddresses)
+	inmemoryBlocks             = 100 // Number of blocks to precompute validators' addresses
+	inmemoryValidatorsPerBlock = 30  // Approximate number of validators' addresses from ecrecover
+	recentAddresses, _         = lru.NewARC(inmemoryBlocks)
+	signatureAddresses, _      = lru.New(inmemoryBlocks * inmemoryValidatorsPerBlock)
```

Comment:
How about combing them into one cache and set/get items insides of `GetSignatureAddress`?

Reply:
I think combining is fine, but the `GetSignatureAddress` is an util function, so putting the cache inside is not good. What do you think?

Reply:
It makes sense. You might create a wrapper function such as `CacheSignatureAddress` or put the caching logic outside as it was.

Reply:
I have added `cacheSignatureAddress` for caching. PTAL.